lib/Sema/CSBindings.cpp

```diff
@@ -1071,6 +1071,15 @@ bool BindingSet::favoredOverConjunction(Constraint *conjunction) const {
     if (forClosureResult() || forGenericParameter())
       return false;
   }
+
+  if (conjunction->getLocator()->isForSingleValueStmtConjunction()) {
+    // If the bindings should be delayed, try the conjunction first.
+    // TODO: Seems like this ought to apply generally? Currently regresses a
+    // diagnostic in a test.
+    if (isDelayed())
+      return false;
+  }
```

Comment:
@xedin Would this make sense to apply more generally? I haven't had a chance to dig into the diagnostic regression. The case it's handling for if/switch expressions is if they're in an array literal, this stops us eagerly binding `[Any]` without solving the conjunction. That being said, that's currently a case we don't allow, so I think we could totally rip this out for now.

Reply:
Do we still need this? I think this conjunction should behave just like a regular one?

Reply:
The need for it is still present, at least if we ever allow if/switch expressions in array literals. That being said, I'm fine with removing for now given that we don't yet allow them there (we'll just emit an, extra, spurious type error).

The reason this hasn't been needed before is that closure solving is triggered by type variable binding, not conjunction solving, so this isn't handled by the existing delaying logic.

Reply:
I'd say let's remove it and cross that bridge when we get there, it's possible that we might work out a more general solution before that point.